### PR TITLE
feat(auth/github): add organization membership check to GitHub

### LINKED ISSUE #2065: [FLI-710] Add organization membership check to GitHub authentication method
Originally requested in Discord https://discord.com/channels/960634591000014878/1146699437536727041

We should add a feature to limit who can authenticate using the GitHub authentication method, based on their organization memberships.

One simple approach we could take here is to support a configuration field identifying the desired organization membership (call it `organization` or `member_of` or something).
This would be similar to the `email_matches` property on the OIDC authentication method.

Then we can use the access token retrieved to get the list of organizations the user is a member of (`api.github.com/users/<user>/orgs`). In an authenticated context, this should return the users organizations. Though it _may_ need the additional scope `read:org` (I think team membership definitely needs this, but just listing org membership might be fine with the `read:user` scope).
Once we have their associated orgs, we could ensure that the lists contains the configured organization name.

This would likely require the following changes (for anyone interested in giving this a go):
- List org and enforcement check [here](https://github.com/flipt-io/flipt/blob/main/internal/server/auth/method/github/server.go#L161). This would be conditional, based on whether or not an organization had been configured. Plus unit test.
- New configuration option on [here](https://github.com/flipt-io/flipt/blob/main/internal/config/authentication.go#L423-L430). Probably best to just call it `organization`.
- Updates to both the CUE and JSON Schema to reflect the new configuration option.

<sub>[FLI-710](https://linear.app/flipt/issue/FLI-710/add-organization-membership-check-to-github-authentication-method)</sub>

### PR DESCRIPTION
**Pull request by erka**
add organization membership check to GitHub authentication method

fixes #2065

### CODE REVIEW

**Comment by erka** [comment]
@markphelps I believe there could be something missing. It requires `read:org` scope to get user's organizations. Maybe there should be some configuration check but it's an open question for me. Please take a lot and provide the feedback.

**Comment by markphelps** [comment]
> @markphelps I believe there could be something missing. It requires `read:org` scope to get user's organizations. Maybe there should be some configuration check but it's an open question for me. Please take a lot and provide the feedback.

Cheers @erka ! Thanks for taking this on! I have a proposed solution to ^, https://github.com/erka/flipt/pull/4

I still need to add tests however, but can do that in the fork PR or this one

File: internal/config/authentication.go [hunk 1]
```diff
@@ -429,10 +429,11 @@ func (a AuthenticationMethodKubernetesConfig) info() AuthenticationMethodInfo {
...
 type AuthenticationMethodGithubConfig struct {
-	ClientId        string   `json:"-" mapstructure:"client_id" yaml:"-"`
-	ClientSecret    string   `json:"-" mapstructure:"client_secret" yaml:"-"`
-	RedirectAddress string   `json:"redirectAddress,omitempty" mapstructure:"redirect_address" yaml:"redirect_address,omitempty"`
-	Scopes          []string `json:"scopes,omitempty" mapstructure:"scopes" yaml:"scopes,omitempty"`
+	ClientId             string   `json:"-" mapstructure:"client_id" yaml:"-"`
+	ClientSecret         string   `json:"-" mapstructure:"client_secret" yaml:"-"`
+	RedirectAddress      string   `json:"redirectAddress,omitempty" mapstructure:"redirect_address" yaml:"redirect_address,omitempty"`
+	Scopes               []string `json:"scopes,omitempty" mapstructure:"scopes" yaml:"scopes,omitempty"`
+	AllowedOrganizations []string `json:"allowed_organizations,omitempty" mapstructure:"allowed_organizations" yaml:"allowed_organizations,omitempty"`
```
Comment (markphelps) [inline internal/config/authentication.go:436]:
```suggestion
	AllowedOrganizations []string `json:"allowedOrganizations,omitempty" mapstructure:"allowed_organizations" yaml:"allowed_organizations,omitempty"`
```

This has been a thorn in our side for a while where the JSON is camel case, but all the other struct tags are snake case.

File: internal/server/auth/method/github/server.go [hunk 2]
```diff
@@ -185,3 +180,36 @@ func (s *Server) Callback(ctx context.Context, r *auth.CallbackRequest) (*auth.C
 		Authentication: a,
 	}, nil
 }
+
+type githubSimpleOrganization struct {
+	Login string
+}
+
+// api calls Github API, decodes and stores successful response in the value pointed to by v.
+func (s *Server) api(ctx context.Context, token *oauth2.Token, endpoint endpoint, v any) error {
```
Comment (markphelps) [inline internal/server/auth/method/github/server.go:189]:
```suggestion
func api(ctx context.Context, token *oauth2.Token, endpoint endpoint, v any) error {
```

i don't think this need to be a method as it doesn't using anything on `s` it seems. could just be a `func` I think

**Review by markphelps — COMMENTED** [review]
looking great, couple suggestions

File: internal/server/auth/method/github/server.go (hunk 2 above)
Comment (erka) [reply internal/server/auth/method/github/server.go]: clean it up

**Review by markphelps — APPROVED** [review]
Looks and works great!

CleanShot 2023-12-10 at 14 08 05

I tried it out by setting up our test OAuth app on Github

When I configured it to check:
`allowed_organizations: flipt-io`

It let me in. But when I changed it to `allowed_organizations: flip-io` purposefully misspelled, it gave me a 401 response!

CleanShot 2023-12-10 at 14 10 56

**Comment by markphelps** [comment]
also checked that it errors if you don't request the `read:org` scope when setting `allowed_organizations` !

```
Error: loading configuration scopes must contain read:org when allowed_organizations is not empty
```

thanks @erka ! I will work on docs updates and get this out probably tomorrow in a release!

**Pull request merged**